### DISCUSSION TITLE
Why does DomainTestModule depends on EntityFrameworkCoreTestModule?

### DISCUSSION THREAD

**Discussion by jflaga** [abpframework/abp#17469] [Q&A]
I have a question about the generated `DomainTestModule` for ABP.
There is a comment at the top saying "... however your domain tests should be database independent anyway", 

```

/* Domain tests are configured to use the EF Core provider.
 * You can switch to MongoDB, however your domain tests should be
 * database independent anyway.
 */
[DependsOn(
    typeof(SampleEntityFrameworkCoreTestModule)
    )]
public class SampleDomainTestModule : AbpModule
{

}
```

but `DomainTestModule` depends on `EntityFrameworkCoreTestModule`.

Is this a mistake? \
If it's not a mistake, is it okay, or even recommended, to remove the `DomainTestModule`'s dependency on `EntityFrameworkCoreTestModule`? \
If it's not recommended to remove it, what are the reason(s) why `DomainTestModule` depends on `EntityFrameworkCoreTestModule`?

Thanks

**Comment by maliming** [comment]
hi

Can you create an issue?

We will abstract all db-related tests in the domain. then implement these in ef and mongo test projects.

**Reply by jflaga** [reply]
Hi @maliming , I created the issue here: https://github.com/abpframework/abp/issues/17488

**Reply by jflaga** [reply]
Hi @maliming , I've been checking out the test modules in the latest version of ABP (v 8.3). And I have come to realize that the `DomainTestModule` is for Integration Tests, not Unit Tests.

When I asked the question in the original post above, I did not yet understand the purpose of `DomainTestModule`. I was assuming that `DomainTestModule` is for Unit Tests of the domain module. I thought that the comment _"domain tests should be database independent"_ on the top of the class `SampleDomainTestModule` is an indication that this module is for unit tests, because unit tests are database independent.

That's the reason why I asked whether I could remove the reference for `EntityFrameworkCoreTestModule`, because that reference is not needed when writing unit tests for the domain module.

Now that I understand the purpose of `DomainTestModule` and of the `.Domain.Tests` project, I think the original structure of the test projects (where `.Domain.Tests` depends on `.EntityFrameworkCore.Tests`) is better than the newer structure  where `.EntityFrameworkCore.Tests` depends on `.Domain.Tests`), because the original structure is simpler, and in the original structure we can easily debug individual tests in Visual Studio by right-clicking on a test function in the code editor.

But I have a structure in mind which, in my opinion, is a much better structure for the tests:

Remove the `.Application.Tests` and `.Domain.Tests` projects. And then rename `EntityFrameworkCore.Tests` project into `.IntegrationTests` . Then put all the sample integration tests for `AppServices` and `DomainServices` in the `.IntegrationTests` project.

We can also keep the domain test project, but repurpose it to become a Unit Test project by renaming it into `.Domain.UnitTests`, and removing the dependency on `EntityFrameworkCoreTestModule` (make it depend on the `.Domain` project only)

<pre>
src
  - ...
  - Sample.Domain
  - ...
test
  - Sample.Domain.UnitTests
  - Sample.IntegrationTests
  - Sample.IntegrationTestBase
  - Sample.Web.Tests
</pre>

That is just a suggestion.

Thanks for your work.
